src/init.cpp

```diff
@@ -1645,8 +1645,8 @@ bool AppInitMain()
     // Wait for genesis block to be processed
     {
         WaitableLock lock(cs_GenesisWait);
-        while (!fHaveGenesis) {
-            condvar_GenesisWait.wait(lock);
+        while (!fHaveGenesis && !ShutdownRequested()) {
+            condvar_GenesisWait.wait_for(lock, std::chrono::milliseconds(500));
```

Comment:
In commit "Fix fast-shutdown hang on ThreadImport+GenesisWait"

Can you add comment explaining why timeout is needed here and wait() is not sufficient?

Reply:
So this means that if ShutdownRequested() is set it will fall out of this loop, without a genesis block. Don't we also have to make sure that AppInitMain() immediately exits (with false) in that case? Looks like an error waiting to happen if it proceeds the initialization without genesis block.